File: python/bigasterisk/testgen.py
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Systematic test-input generation for Spark SQL.

Fuzzing searches for inputs by mutating what it has. This works the other way round: it
reads the query's own conditions, solves them, and constructs one input per path through
them. Where fuzzing eventually stumbles onto a branch, this builds a record that takes
it.

This is BigTest's approach; with ``natural=True`` (the default) it is NaturalSym's — the
same paths, but with witnesses drawn from values that really occur, so the generated
tests read like data rather than like solver output.

    import bigasterisk

    suite = bigasterisk.testgen(spark).generate(
        "SELECT cid FROM orders WHERE amount > 100", {"orders": orders})

    print(suite.coverage)
    for case in suite.cases:
        print(case)

Every generated input is executed and the branch it was built for is checked, so
``case.verified`` says what actually happened rather than what was intended.
"""

import json


class TestCase:
    """An input built to drive the query down a particular path."""

    def __init__(self, payload):
        self.id = payload["id"]
        self.path = payload["path"]
        self.verified = payload["verified"]
        self.note = payload["note"]
        self.tables = payload["tables"]

    def __repr__(self):
        mark = "ok" if self.verified else "--"
        rows = "\n".join(
            "    %s: %s" % (name, ", ".join(rows)) for name, rows in self.tables.items())
        return "[%s] %s  (%s)\n%s" % (mark, self.path, self.note, rows)


class TestSuite:
    """What a test-generation run produced."""

    def __init__(self, payload):
        self.total_branches = payload["totalBranches"]
        self.coverage = payload["coverage"]
        self.cases = [TestCase(c) for c in payload["cases"]]

    @property
    def verified(self):
        """Tests whose path was reached when the input was actually run."""
        return [c for c in self.cases if c.verified]

    def __repr__(self):
        return "TestSuite(%d cases, %d verified, %d branches)" % (
            len(self.cases), len(self.verified), self.total_branches)


class TestGen:
    """Systematic test-input generation for a SparkSession."""

    def __init__(self, spark):
        self._spark = spark
        self._support = spark._jvm.org.bigasterisk.api.BigAsterisk.testgen(
            spark._jsparkSession)

    def generate(self, query, seeds, max_paths=32, rows_per_path=3, natural=True,
                 seed=0, distributions=None):
        """Generate a test suite for ``query``.

        ``seeds`` maps each table name the query reads to a DataFrame. Schemas are
        required; the rows are the pool of natural witnesses.

        ``distributions`` declares the shape of individual columns by name, as text::

            distributions={"score": "binom(100, 0.1)",
                           "name": 'Discrete("alice", "bob")'}

        You know the shape of your own data; the solver does not, and left to itself
        satisfies ``age > 18`` with ``19`` every time. Where a column is declared,
        witnesses are drawn from that distribution and kept if they satisfy the path —
        so naturalness never costs coverage.

        Generation swaps its inputs in under those table names while it runs and
        restores the originals afterwards.
        """
        jseeds = self._spark._jvm.java.util.HashMap()
        for name, df in seeds.items():
            jseeds.put(name, df._jdf)
        jdists = self._spark._jvm.java.util.HashMap()
        for column, spec in (distributions or {}).items():
            jdists.put(column, spec)
        suite = self._support.generateJava(
            query, jseeds, int(max_paths), int(rows_per_path), bool(natural), int(seed),
            jdists)
        return TestSuite(json.loads(suite.json()))

File: modules/api/src/main/scala/org/bigasterisk/api/TestGen.scala
package org.bigasterisk.api

import org.apache.spark.sql.{DataFrame, Row}

/**
 * How a test-generation run is configured.
 *
 * @param maxPaths  the most paths to enumerate. Paths grow exponentially with the
 *                  number of branches, so this is a real bound, not a formality.
 * @param rowsPerPath rows to generate per table for each path
 * @param natural   when true, witnesses are drawn from values that actually occur in
 *                  the seed data wherever a constraint allows it, so generated tests
 *                  look like real records rather than like solver output. This is what
 *                  separates NaturalSym from BigTest.
 * @param seed      random seed, so a run reproduces exactly
 *
 * @group testgen
 */
case class TestGenConfig(
    maxPaths: Int = 32,
    rowsPerPath: Int = 3,
    natural: Boolean = true,
    seed: Long = 0L) {
  require(maxPaths > 0, s"maxPaths must be positive, got $maxPaths")
  require(rowsPerPath > 0, s"rowsPerPath must be positive, got $rowsPerPath")
}

/**
 * One generated test: an input built to drive the query down a particular path.
 *
 * @param id       position in the generated suite
 * @param path     the branch outcomes this input was built to take, as SQL text
 * @param tables   the generated rows, by table name
 * @param verified whether running the query with these rows actually took that path.
 *                 A generator that reports coverage it did not achieve is worse than
 *                 useless, so every test is executed and checked.
 * @param note     what happened — `"verified"`, `"unsatisfiable"`, or why the path was
 *                 not reached
 *
 * @group testgen
 */
case class TestCase(
    id: Int,
    path: String,
    tables: Map[String, Seq[Row]],
    verified: Boolean,
    note: String) {

  override def toString: String = {
    val mark = if (verified) "ok" else "--"
    val rows = tables.map { case (t, r) => s"    $t: ${r.mkString(", ")}" }.mkString("\n")
    s"[$mark] $path  ($note)\n$rows"
  }
}

/**
 * What a test-generation run produced.
 *
 * @group testgen
 */
case class TestSuite(cases: Seq[TestCase], totalBranches: Int) {

  /** Tests whose path was reached when the input was actually run. */
  def verified: Seq[TestCase] = cases.filter(_.verified)

  /** Branches some verified test reached. */
  def coveredBranches: Set[String] =
    verified.flatMap(_.path.split(" AND ").map(_.trim)).toSet

  /** Fraction of the query's branches a verified test reaches. */
  def coverage: Double =
    if (totalBranches == 0) 1.0
    else math.min(1.0, coveredBranches.count(!_.startsWith("NOT ")).toDouble / totalBranches)

  override def toString: String =
    s"TestSuite(${cases.size} cases, ${verified.size} verified, " +
      s"$totalBranches branches)"

  /**
   * The suite as a JSON object.
   *
   * Provided for language bindings that cannot marshal Scala collections across the
   * process boundary — the PySpark front end reads this. Generated rows are rendered
   * with their `toString`, which is what a test case needs to be readable.
   */
  def json: String = {
    def quote(s: String): String =
      "\"" + s.flatMap {
        case '"'              => "\\\""
        case '\\'             => "\\\\"
        case '\n'             => "\\n"
        case '\r'             => "\\r"
        case '\t'             => "\\t"
        case c if c.isControl => f"\\u${c.toInt}%04x"
        case c                => c.toString
      } + "\""

    val casesJson = cases.map { c =>
      val tables = c.tables.map { case (name, rows) =>
        s"${quote(name)}:[${rows.map(r => quote(r.toString)).mkString(",")}]"
      }.mkString(",")
      s"""{"id":${c.id},"path":${quote(c.path)},"verified":${c.verified},""" +
        s""""note":${quote(c.note)},"tables":{$tables}}"""
    }.mkString(",")

    s"""{"totalBranches":$totalBranches,"coverage":$coverage,"cases":[$casesJson]}"""
  }
}

/**
 * Systematic test-input generation for Spark SQL.
 *
 * Fuzzing searches for inputs by mutating what it has. Symbolic test generation works
 * the other way round: it reads the query's own conditions, solves them, and produces
 * one input per path through them. Where fuzzing eventually stumbles onto a branch,
 * this constructs a record that takes it.
 *
 * This is BigTest's approach (ESEC/FSE 2019), applied to SQL predicates. With
 * `natural = true` it is NaturalSym's (FSE 2024): the same paths, but with witnesses
 * drawn from values that really occur, so the generated tests read like data.
 *
 * Obtain an instance from [[BigAsterisk.testgen]].
 *
 * {{{
 * val suite = BigAsterisk.testgen(spark).generate(
 *   "SELECT cid FROM orders WHERE amount > 100",
 *   Map("orders" -> orders))
 *
 * suite.coverage
 * suite.cases.foreach(println)
 * }}}
 *
 * @group testgen
 */
trait TestGenSupport {

  /**
   * Generates a test suite for `query`.
   *
   * @param seeds the tables the query reads, by the name it reads them under. Schemas
   *        are required; rows are used as the pool of natural witnesses.
   */
  def generate(query: String, seeds: Map[String, DataFrame], config: TestGenConfig): TestSuite

  /** Generates a suite with the default configuration. */
  final def generate(query: String, seeds: Map[String, DataFrame]): TestSuite =
    generate(query, seeds, TestGenConfig())

  /**
   * [[generate]] for callers that cannot build a Scala `Map` or a [[TestGenConfig]] —
   * Java, and Py4J, which marshals a Python dict to `java.util.Map`.
   */
  final def generateJava(
      query: String,
      seeds: java.util.Map[String, DataFrame],
      maxPaths: Int,
      rowsPerPath: Int,
      natural: Boolean,
      seed: Long): TestSuite = {
    import scala.jdk.CollectionConverters._
    generate(query, seeds.asScala.toMap,
      TestGenConfig(maxPaths, rowsPerPath, natural, seed))
  }
}
